drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+                    WriteState::ResponseWritten
+                    | WriteState::Flush
+                    | WriteState::Underrun => {
+                        return;
+                    }
+                }
+            }
         }
+    }
 
-        // TODO: SP RESET needs to be monitored, otherwise, any
-        // forced looping here could be a denial of service attack against
-        // observation of SP resetting. SP resetting without invalidating
-        // security related state means a compromised SP could operate using
-        // the trust gained in the previous session.
-        // Upper layers may mitigate that, but check on it.
-
-        // Wait for chip select to be asserted and perform all subsequent I/O
-        // for that frame.
-        // Track the state of chip select (CSn)
-        let mut inframe = false;
-        'outer: loop {
-            // restart here on the one expected spurious interrupt.
+    /// Read a request from the SP
+    ///
+    /// We clock out 0s until CSn is de-asserted
+    pub fn spi_read<'a>(
+        &mut self,
+        mut rx_msg: RxMsg<'a>,
+        tx_msg: TxMsg<'a>,
+        handler: &mut Handler,
+    ) -> Option<VerifiedTxMsg<'a>> {
+        self.state = State::Read(ReadState::WaitingForRequest);
+        self.zero_tx_buf();
+        loop {
             sys_irq_control(SPI_IRQ, true);
             sys_recv_closed(&mut [], SPI_IRQ, TaskId::KERNEL).unwrap_lite();
             loop {
-                // Get frame start/end interrupt from intstat (SSA/SSD).
-                let intstat = self.spi.intstat();
-                let fifostat = self.spi.fifostat();
-
-                // During bulk data transfer, we'll be polling and not servicing
-                // the interrupts that the kernel is collecting.
-                // As a result, there can be a left-over kernel interrupt to handle
-                // even after the HW interrupts are retired.
-                if !intstat.ssa().bit()
-                    && !intstat.ssd().bit()
-                    && !fifostat.txnotfull().bit()
-                    && !fifostat.rxnotempty().bit()
-                    && !inframe
-                {
-                    // This is a spurious interrupt.
-                    // These are only happening just after queuing a
-                    // response.
-                    // TODO: It would be nice to eliminate the spurious interrupt.
-                    continue 'outer;
+                ringbuf_entry!(Trace::State(self.state));
+                match self.state.read() {
+                    ReadState::WaitingForRequest => {
+                        self.read_wait_for_csn_asserted();
+                        if self.state
+                            == State::Read(ReadState::WaitingForRequest)
+                        {
+                            // Wait for interrupt
+                            break;
+                        }
```

Comment:
If there is a valid interrupt, then we should be in the tight I/O loop as soon as possible.
The rest of the state machine is not critical, but I would inline the interrupt test and tight loop instead of letting the compiler think that anything could happen next.

Reply:
The attempt with this code was to only interrupt on CSn asserted, and that gets checked for in `self.read_wait_for_csn_asserted()`. If the state doesn't change (`self.state == State::Read(ReadState::WaitingForRequest)`) then CSn wasn't asserted. 

I'm not even sure why the interrupt would fire if CSn wasn't asserted. Would an interrupt firing for any other reason be "valid" in this case?

Reply:
Code no longer exists.